### PR TITLE
Apply variable-spaced optimization to QROM circuits

### PR DESCRIPTION
**Pull request by tanujkhattar**
Updates the unary iteration framework to allow "early termination" and uses this trick to apply the variable-spacing optimization to QROM circuits described in https://arxiv.org/pdf/2007.07391.pdf

As a drive-by fix, the PR also fixes a flaky bug in `GreedyQubitManager` where the free'd qubits were added to the pool of managed qubits in a non-deterministic order due to finding unique elements via `set(qubits)`. Replaced with an order-preserving way of finding unique elements.

### CODE REVIEW

File: cirq-ft/cirq_ft/algos/qrom.py [hunk 1]
```diff
@@ -152,11 +165,23 @@ def decompose_zero_selection(
             yield cirq.inverse(multi_controlled_and)
             context.qubit_manager.qfree(and_ancilla + [and_target])
 
+    def _break_early(self, selection_index_prefix: Tuple[int, ...], l: int, r: int):
+        global_unique_element = set()
+        for data in self.data:
+            unique_element = np.unique(data[selection_index_prefix][l:r])
+            print(selection_index_prefix, l, r, data[selection_index_prefix], unique_element)
```
Comment (fdmalone) [inline cirq-ft/cirq_ft/algos/qrom.py:172]: Remove print statement?

File: cirq-ft/cirq_ft/algos/unary_iteration_gate.py [hunk 2]
```diff
@@ -308,10 +337,39 @@ def decompose_zero_selection(
...
+        For each internal node of the unary iteration segment tree, `break_early(l, r)` is called
+        to evaluate whether the unary iteration should not recurse in the subtree of the node
+        representing range `[l, r)`. If True, the internal node is considered equivalent to a leaf
+        node and thus, `self.nth_operation` will be called for only integer `l` in the range [l, r).
+
+        When the `UnaryIteration` class is constructed using multiple selection registers, i.e. we
+        wish to perform nested coherent for-loops, a unary iteration segment tree is constructed
+        corresponding to each nested coherent for-loop. For every such unary iteration segment tree,
+        the `_break_early` condition is checked by passing the `selection_index_prefix` tuple. It is
+        equivalent to the following classical code.
```
Comment (fdmalone) [inline cirq-ft/cirq_ft/algos/unary_iteration_gate.py:352]: What following classical code?

File: cirq-ft/cirq_ft/algos/qrom.py (hunk 1 above)
Comment (tanujkhattar) [reply cirq-ft/cirq_ft/algos/qrom.py]: done

File: cirq-ft/cirq_ft/algos/unary_iteration_gate.py (hunk 2 above)
Comment (tanujkhattar) [reply cirq-ft/cirq_ft/algos/unary_iteration_gate.py]: removed the last line

File: cirq-ft/cirq_ft/algos/qrom_test.py [hunk 3]
```diff
@@ -116,6 +116,76 @@ def test_t_complexity(data):
...
+    anc = sorted(set(circuit.all_qubits()) - set(op.qubits))
+    selection_and_anc = (selection[0],) + sum(zip(selection[1:], anc), ())
+    qubit_order = cirq.QubitOrder.explicit(selection_and_anc, fallback=cirq.QubitOrder.DEFAULT)
+    cirq.testing.assert_has_diagram(circuit, expected_diagram, qubit_order=qubit_order)
+
+
+def test_qrom_variable_spacing():
+    # Tests for variable spacing optimization applied from https://arxiv.org/abs/2007.07391
+    data = [1, 2, 3, 4, 5, 5, 6, 6, 7, 7, 7, 7, 8, 8, 8, 8]  # Figure 3a.
+    assert cirq_ft.t_complexity(cirq_ft.QROM.build(data)).t == 6 * 4
```
Comment (fdmalone) [inline cirq-ft/cirq_ft/algos/qrom_test.py:137]: Is it possible to add a test for the negative case they give in the paper. As in, " For example, we can use
a region such as {4, 5}, but not {3, 4}, because 4 ≡ 100 and 5 ≡ 101, so grouping 4 and 5 corresponds to ignoring the
least significant bit, but the least significant bit changes between 3 and 4."

**Review by fdmalone — CHANGES_REQUESTED** [review]
LGTM % adding an additional test case.

File: cirq-ft/cirq_ft/algos/qrom_test.py (hunk 3 above)
Comment (tanujkhattar) [reply cirq-ft/cirq_ft/algos/qrom_test.py]: added the test

**Comment by tanujkhattar** [comment]
@fdmalone Added more tests and also updated the `_value_equality_values_` method because the T-complexity will now depend upon the data to be loaded and not just on the shape of the registers.

**Review by fdmalone — COMMENTED** [review]
LGTM

**Review by fdmalone — APPROVED** [review]
